## Covid Data Exploration and Analysis 
- Analyzing covid deaths and vaccinations for whole world and India
- Data Source -  ourworldindata
- finding highest death counts
- finding percatnges of deaths and total vaccinations with respect to population
- Analysis of data on country and continent level

#### Steps
- reading the data
- Cleaning for missing values
- seperating data on country, continents level
- finding useful dates for analysis for vaccinations


In [1]:
import pandas as pd
import pandasql as ps

In [2]:
cd = pd.read_excel("https://github.com/Siddharthbadal/SQL/blob/main/dataset/covidDeaths.xlsx?raw=true", sheet_name="covidDeaths")
cd.head()

iso_code continent     location       date  population  total_cases  \
0      AFG      Asia  Afghanistan 2020-02-24  38928341.0          1.0   
1      AFG      Asia  Afghanistan 2020-02-25  38928341.0          1.0   
2      AFG      Asia  Afghanistan 2020-02-26  38928341.0          1.0   
3      AFG      Asia  Afghanistan 2020-02-27  38928341.0          1.0   
4      AFG      Asia  Afghanistan 2020-02-28  38928341.0          1.0   

   new_cases  new_cases_smoothed  total_deaths  new_deaths  ...  \
0        1.0                 NaN           NaN         NaN  ...   
1        0.0                 NaN           NaN         NaN  ...   
2        0.0                 NaN           NaN         NaN  ...   
3        0.0                 NaN           NaN         NaN  ...   
4        0.0                 NaN           NaN         NaN  ...   

   new_deaths_smoothed_per_million  reproduction_rate  icu_patients  \
0                              NaN                NaN           NaN   
1                              NaN                NaN           NaN   
2                              NaN                NaN           NaN   
3                              NaN                NaN           NaN   
4                              NaN                NaN           NaN   

   icu_patients_per_million  hosp_patients  hosp_patients_per_million  \
0                       NaN            NaN                        NaN   
1                       NaN            NaN                        NaN   
2                       NaN            NaN                        NaN   
3                       NaN            NaN                        NaN   
4                       NaN            NaN                        NaN   

   weekly_icu_admissions  weekly_icu_admissions_per_million  \
0                    NaN                                NaN   
1                    NaN                                NaN   
2                    NaN                                NaN   
3                    NaN                                NaN   
4                    NaN                                NaN   

   weekly_hosp_admissions  weekly_hosp_admissions_per_million  
0                     NaN                                 NaN  
1                     NaN                                 NaN  
2                     NaN                                 NaN  
3                     NaN                                 NaN  
4                     NaN                                 NaN  

[5 rows x 26 columns]

In [3]:
cv = pd.read_excel('https://github.com/Siddharthbadal/SQL/blob/main/dataset/covidVaccines.xlsx?raw=true', sheet_name="covidVaccines")
cv.head()

iso_code continent     location       date  new_tests  total_tests  \
0      AFG      Asia  Afghanistan 2020-02-24        NaN          NaN   
1      AFG      Asia  Afghanistan 2020-02-25        NaN          NaN   
2      AFG      Asia  Afghanistan 2020-02-26        NaN          NaN   
3      AFG      Asia  Afghanistan 2020-02-27        NaN          NaN   
4      AFG      Asia  Afghanistan 2020-02-28        NaN          NaN   

   total_tests_per_thousand  new_tests_per_thousand  new_tests_smoothed  \
0                       NaN                     NaN                 NaN   
1                       NaN                     NaN                 NaN   
2                       NaN                     NaN                 NaN   
3                       NaN                     NaN                 NaN   
4                       NaN                     NaN                 NaN   

   new_tests_smoothed_per_thousand  ...  extreme_poverty  \
0                              NaN  ...              NaN   
1                              NaN  ...              NaN   
2                              NaN  ...              NaN   
3                              NaN  ...              NaN   
4                              NaN  ...              NaN   

   cardiovasc_death_rate diabetes_prevalence  female_smokers  male_smokers  \
0                597.029                9.59             NaN           NaN   
1                597.029                9.59             NaN           NaN   
2                597.029                9.59             NaN           NaN   
3                597.029                9.59             NaN           NaN   
4                597.029                9.59             NaN           NaN   

   handwashing_facilities  hospital_beds_per_thousand  life_expectancy  \
0                  37.746                         0.5            64.83   
1                  37.746                         0.5            64.83   
2                  37.746                         0.5            64.83   
3                  37.746                         0.5            64.83   
4                  37.746                         0.5            64.83   

   human_development_index  excess_mortality  
0                    0.511               NaN  
1                    0.511               NaN  
2                    0.511               NaN  
3                    0.511               NaN  
4                    0.511               NaN  

[5 rows x 38 columns]

## Selecting data we need for analysis

In [4]:
ps.sqldf("""

SELECT location, date, total_cases, total_deaths
FROM cd
ORDER BY 1, 2
LIMIT 10
""")

location                        date  total_cases total_deaths
0  Afghanistan  2020-02-24 00:00:00.000000          1.0         None
1  Afghanistan  2020-02-25 00:00:00.000000          1.0         None
2  Afghanistan  2020-02-26 00:00:00.000000          1.0         None
3  Afghanistan  2020-02-27 00:00:00.000000          1.0         None
4  Afghanistan  2020-02-28 00:00:00.000000          1.0         None
5  Afghanistan  2020-02-29 00:00:00.000000          1.0         None
6  Afghanistan  2020-03-01 00:00:00.000000          1.0         None
7  Afghanistan  2020-03-02 00:00:00.000000          1.0         None
8  Afghanistan  2020-03-03 00:00:00.000000          2.0         None
9  Afghanistan  2020-03-04 00:00:00.000000          4.0         None

In [5]:
ps.sqldf(""" 
   SELECT location, date, total_cases, total_deaths
FROM cd
Where location = 'Africa'
LIMIT 10

""")

location                        date  total_cases total_deaths
0   Africa  2020-02-13 00:00:00.000000          NaN         None
1   Africa  2020-02-14 00:00:00.000000          1.0         None
2   Africa  2020-02-15 00:00:00.000000          1.0         None
3   Africa  2020-02-16 00:00:00.000000          1.0         None
4   Africa  2020-02-17 00:00:00.000000          1.0         None
5   Africa  2020-02-18 00:00:00.000000          1.0         None
6   Africa  2020-02-19 00:00:00.000000          1.0         None
7   Africa  2020-02-20 00:00:00.000000          1.0         None
8   Africa  2020-02-21 00:00:00.000000          1.0         None
9   Africa  2020-02-22 00:00:00.000000          1.0         None

In [6]:
# Deaths Percentages in USA

ps.sqldf("""

SELECT location, date, total_cases, total_deaths, (total_deaths/total_cases)*100 AS "Death %"
FROM cd
WHERE location like "%states%"
ORDER BY 1,2


""")

location                        date  total_cases  total_deaths  \
0    United States  2020-01-22 00:00:00.000000          1.0           NaN   
1    United States  2020-01-23 00:00:00.000000          1.0           NaN   
2    United States  2020-01-24 00:00:00.000000          2.0           NaN   
3    United States  2020-01-25 00:00:00.000000          2.0           NaN   
4    United States  2020-01-26 00:00:00.000000          5.0           NaN   
..             ...                         ...          ...           ...   
518  United States  2021-06-23 00:00:00.000000   33577651.0      602837.0   
519  United States  2021-06-24 00:00:00.000000   33590481.0      603178.0   
520  United States  2021-06-25 00:00:00.000000   33614196.0      603744.0   
521  United States  2021-06-26 00:00:00.000000   33621499.0      603891.0   
522  United States  2021-06-27 00:00:00.000000   33624983.0      603966.0   

      Death %  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  
..        ...  
518  1.795352  
519  1.795681  
520  1.796098  
521  1.796145  
522  1.796182  

[523 rows x 5 columns]

In [7]:
# Death % in India. 

ps.sqldf("""

SELECT location, date, total_cases, total_deaths, (total_deaths/total_cases)*100 AS "Death %"
FROM cd
WHERE location = "India"
ORDER BY 1,2


""")

location                        date  total_cases  total_deaths   Death %
0      India  2020-01-30 00:00:00.000000          1.0           NaN       NaN
1      India  2020-01-31 00:00:00.000000          1.0           NaN       NaN
2      India  2020-02-01 00:00:00.000000          1.0           NaN       NaN
3      India  2020-02-02 00:00:00.000000          2.0           NaN       NaN
4      India  2020-02-03 00:00:00.000000          3.0           NaN       NaN
..       ...                         ...          ...           ...       ...
510    India  2021-06-23 00:00:00.000000   30082778.0      391981.0  1.303008
511    India  2021-06-24 00:00:00.000000   30134445.0      393310.0  1.305184
512    India  2021-06-25 00:00:00.000000   30183143.0      394493.0  1.306998
513    India  2021-06-26 00:00:00.000000   30233183.0      395751.0  1.308995
514    India  2021-06-27 00:00:00.000000   30279331.0      396730.0  1.310234

[515 rows x 5 columns]

## Understanding continent column in the table
Continents with the Hihgest deaths count

In [8]:
ps.sqldf("""
    SELECT continent, MAX(Total_deaths) AS "Total Deaths In Continents"
    FROM cd
    WHERE continent is not null
    GROUP By continent
    ORDER By "Total Deaths In Continents" DESC

""")

continent  Total Deaths In Continents
0  North America                    603966.0
1  South America                    513474.0
2           Asia                    396730.0
3         Europe                    131070.0
4         Africa                     59900.0
5        Oceania                       910.0

In [9]:
# Data where continent is null
ps.sqldf("""
    SELECT continent, MAX(Total_deaths) AS "Total Deaths In Continents"
    FROM cd
    WHERE continent is null
    GROUP By continent
    ORDER By "Total Deaths In Continents" DESC

""")

continent  Total Deaths In Continents
0      None                   3923066.0

## Using location instead of Continent
Countries with highest death count where continent values are not missing

In [10]:
ps.sqldf("""
    SELECT location, MAX(Total_deaths) AS "Total Deaths In Continents"
    FROM cd
    WHERE continent is not null
    GROUP By location
    ORDER By "Total Deaths In Continents" DESC

""")

location  Total Deaths In Continents
0                      United States                    603966.0
1                             Brazil                    513474.0
2                              India                    396730.0
3                             Mexico                    232564.0
4                               Peru                    191584.0
..                               ...                         ...
216           British Virgin Islands                         NaN
217  Bonaire Sint Eustatius and Saba                         NaN
218                          Bermuda                         NaN
219                            Aruba                         NaN
220                         Anguilla                         NaN

[221 rows x 2 columns]

#### now considering null valuess in continent with location. This gives more accurate results

In [11]:

ps.sqldf("""
    SELECT location, MAX(Total_deaths) AS "Total Deaths In Continents"
    FROM cd
    WHERE continent is null
    GROUP By location
    ORDER By "Total Deaths In Continents" DESC

""")

location  Total Deaths In Continents
0           World                   3923066.0
1          Europe                   1102043.0
2   South America                    994754.0
3   North America                    900444.0
4            Asia                    783567.0
5  European Union                    738307.0
6          Africa                    141118.0
7         Oceania                      1125.0
8   International                        15.0

# Total cases Vs Population

In [12]:
ps.sqldf("""

SELECT location, date, total_cases, population, round((total_cases / population) * 100, 3) as "Death % of Population"
FROM cd
WHERE location like "%states%"
ORDER BY 1,2


""")

location                        date  total_cases   population  \
0    United States  2020-01-22 00:00:00.000000          1.0  331002647.0   
1    United States  2020-01-23 00:00:00.000000          1.0  331002647.0   
2    United States  2020-01-24 00:00:00.000000          2.0  331002647.0   
3    United States  2020-01-25 00:00:00.000000          2.0  331002647.0   
4    United States  2020-01-26 00:00:00.000000          5.0  331002647.0   
..             ...                         ...          ...          ...   
518  United States  2021-06-23 00:00:00.000000   33577651.0  331002647.0   
519  United States  2021-06-24 00:00:00.000000   33590481.0  331002647.0   
520  United States  2021-06-25 00:00:00.000000   33614196.0  331002647.0   
521  United States  2021-06-26 00:00:00.000000   33621499.0  331002647.0   
522  United States  2021-06-27 00:00:00.000000   33624983.0  331002647.0   

     Death % of Population  
0                    0.000  
1                    0.000  
2                    0.000  
3                    0.000  
4                    0.000  
..                     ...  
518                 10.144  
519                 10.148  
520                 10.155  
521                 10.157  
522                 10.159  

[523 rows x 5 columns]

In [13]:
ps.sqldf("""

SELECT location, date, total_cases, population, round((total_cases / population) * 100,2) as "Death % of Population"
FROM cd
WHERE location = "India"
ORDER BY 1,2


""")

location                        date  total_cases    population  \
0      India  2020-01-30 00:00:00.000000          1.0  1.380004e+09   
1      India  2020-01-31 00:00:00.000000          1.0  1.380004e+09   
2      India  2020-02-01 00:00:00.000000          1.0  1.380004e+09   
3      India  2020-02-02 00:00:00.000000          2.0  1.380004e+09   
4      India  2020-02-03 00:00:00.000000          3.0  1.380004e+09   
..       ...                         ...          ...           ...   
510    India  2021-06-23 00:00:00.000000   30082778.0  1.380004e+09   
511    India  2021-06-24 00:00:00.000000   30134445.0  1.380004e+09   
512    India  2021-06-25 00:00:00.000000   30183143.0  1.380004e+09   
513    India  2021-06-26 00:00:00.000000   30233183.0  1.380004e+09   
514    India  2021-06-27 00:00:00.000000   30279331.0  1.380004e+09   

     Death % of Population  
0                     0.00  
1                     0.00  
2                     0.00  
3                     0.00  
4                     0.00  
..                     ...  
510                   2.18  
511                   2.18  
512                   2.19  
513                   2.19  
514                   2.19  

[515 rows x 5 columns]

## Highest Infection rate compared to Population

In [14]:
ps.sqldf("""

SELECT location, population, MAX(total_cases) AS MaxInfectedCount, MAX((total_cases / population)) * 100 AS "InfectedOfPopulation %"
FROM cd
WHERE continent is not null
GROUP BY location, population
ORDER BY "InfectedOfPopulation %" DESC

""")

location  population  MaxInfectedCount  \
0                     Andorra     77265.0           13882.0   
1                  Montenegro    628062.0          100190.0   
2                     Bahrain   1701583.0          265247.0   
3                     Czechia  10708982.0         1666890.0   
4                  Seychelles     98340.0           15116.0   
..                        ...         ...               ...   
216                     Tonga    105697.0               NaN   
217              Turkmenistan   6031187.0               NaN   
218  Turks and Caicos Islands     38718.0               NaN   
219                    Tuvalu     11792.0               NaN   
220         Wallis and Futuna     11246.0               NaN   

     InfectedOfPopulation %  
0                 17.966738  
1                 15.952247  
2                 15.588249  
3                 15.565345  
4                 15.371161  
..                      ...  
216                     NaN  
217                     NaN  
218                     NaN  
219                     NaN  
220                     NaN  

[221 rows x 4 columns]

## Highest Death Rate compared to Population

In [15]:
ps.sqldf("""

SELECT location, population, MAX(total_deaths) AS TotalDeathCount
FROM cd
WHERE continent is not null
GROUP BY location
ORDER BY TotalDeathCount DESC

""")



location    population  TotalDeathCount
0                      United States  3.310026e+08         603966.0
1                             Brazil  2.125594e+08         513474.0
2                              India  1.380004e+09         396730.0
3                             Mexico  1.289328e+08         232564.0
4                               Peru  3.297185e+07         191584.0
..                               ...           ...              ...
216           British Virgin Islands  3.023700e+04              NaN
217  Bonaire Sint Eustatius and Saba  2.622100e+04              NaN
218                          Bermuda  6.227300e+04              NaN
219                            Aruba  1.067660e+05              NaN
220                         Anguilla  1.500200e+04              NaN

[221 rows x 3 columns]

In [16]:
ps.sqldf("""

SELECT location, population, MAX(total_deaths) AS TotalDeathCount, (total_deaths / population) * 100 AS DeathsOfPopulation
FROM cd
WHERE continent is not null
GROUP BY location, population
ORDER BY DeathsOfPopulation DESC

""")


location  population  TotalDeathCount  DeathsOfPopulation
0                        Peru  32971846.0         191584.0            0.581053
1                     Hungary   9660350.0          29980.0            0.310341
2      Bosnia and Herzegovina   3280815.0           9655.0            0.294287
3                     Czechia  10708982.0          30298.0            0.282921
4                  San Marino     33938.0             90.0            0.265189
..                        ...         ...              ...                 ...
216              Turkmenistan   6031187.0              NaN                 NaN
217  Turks and Caicos Islands     38718.0              NaN                 NaN
218                    Tuvalu     11792.0              NaN                 NaN
219                   Vatican       809.0              NaN                 NaN
220         Wallis and Futuna     11246.0              NaN                 NaN

[221 rows x 4 columns]

In [17]:
ps.sqldf("""

SELECT location, total_cases, MAX(total_deaths), MAX((total_deaths / total_cases)) * 100 AS DeathsOnTotalCases
FROM cd
WHERE continent is not null AND
        total_cases > 10000
GROUP BY location, total_cases
ORDER BY DeathsOnTotalCases DESC

""")


location  total_cases  MAX(total_deaths)  DeathsOnTotalCases
0       France      58045.0            13835.0           23.834956
1       France      56600.0            13199.0           23.319788
2       France      55035.0            12214.0           22.193150
3       France      47396.0            10330.0           21.795088
4       France      51251.0            10874.0           21.217147
...        ...          ...                ...                 ...
44289    Qatar      32604.0               15.0            0.046007
44290    Qatar      33969.0               15.0            0.044158
44291    Qatar      38651.0               17.0            0.043983
44292    Qatar      37097.0               16.0            0.043130
44293    Qatar      35606.0               15.0            0.042128

[44294 rows x 4 columns]

### Global Data Analysis
Finding total new deaths with date and
finding total new deaths globally in comparision with total new cases

New cases and new deaths shows data on the given date.

In [18]:
ps.sqldf("""

    SELECT date, SUM(new_cases) AS toal_new_cases, SUM(new_deaths) AS total_new_deaths, SUM(new_deaths * 100/new_cases) AS DeathPercantage
    FROM cd
    WHERE continent is not null AND total_cases is not null
    GROUP BY date
    ORDER by 1, 2


""")

date  toal_new_cases  total_new_deaths  \
0    2020-01-22 00:00:00.000000             NaN               NaN   
1    2020-01-23 00:00:00.000000            98.0               1.0   
2    2020-01-24 00:00:00.000000           286.0               8.0   
3    2020-01-25 00:00:00.000000           492.0              16.0   
4    2020-01-26 00:00:00.000000           685.0              14.0   
..                          ...             ...               ...   
518  2021-06-23 00:00:00.000000        437255.0            9523.0   
519  2021-06-24 00:00:00.000000        403837.0            8612.0   
520  2021-06-25 00:00:00.000000        421740.0            9282.0   
521  2021-06-26 00:00:00.000000        362819.0            7464.0   
522  2021-06-27 00:00:00.000000        309453.0            5948.0   

     DeathPercantage  
0                NaN  
1           1.075269  
2           2.888087  
3           3.312629  
4           2.102102  
..               ...  
518       616.552508  
519       581.593677  
520       605.071707  
521       797.800409  
522       526.109470  

[523 rows x 4 columns]

In [19]:
ps.sqldf("""

    SELECT date, SUM(new_cases) AS toal_new_cases, SUM(new_deaths) AS total_new_deaths, ROUND(SUM(new_deaths/new_cases),2) * 100 AS DeathPercantage
    FROM cd
    WHERE continent is not null AND total_cases is not null
    ORDER by 1, 2


""")

date  toal_new_cases  total_new_deaths  \
0  2020-02-24 00:00:00.000000     180617563.0         3923033.0   

   DeathPercantage  
0         219604.0

### Total deaths are 2.19% of toal cases

##### Joing both the tables covid deaths AS cd and covid vaccinations AS cv

In [5]:
ps.sqldf("""
    SELECT cd.continent, cd.location, cd.date, cd.population, cv.new_vaccinations
    FROM cd
    JOIN cv
    ON cd.location = cv.location AND
    cd.date = cv.date
    WHERE cd.continent is not null
    ORDER BY 1,2,3
""")


continent   location                        date  population  \
0             Africa    Algeria  2020-02-25 00:00:00.000000  43851043.0   
1             Africa    Algeria  2020-02-26 00:00:00.000000  43851043.0   
2             Africa    Algeria  2020-02-27 00:00:00.000000  43851043.0   
3             Africa    Algeria  2020-02-28 00:00:00.000000  43851043.0   
4             Africa    Algeria  2020-02-29 00:00:00.000000  43851043.0   
...              ...        ...                         ...         ...   
94000  South America  Venezuela  2021-06-23 00:00:00.000000  28435943.0   
94001  South America  Venezuela  2021-06-24 00:00:00.000000  28435943.0   
94002  South America  Venezuela  2021-06-25 00:00:00.000000  28435943.0   
94003  South America  Venezuela  2021-06-26 00:00:00.000000  28435943.0   
94004  South America  Venezuela  2021-06-27 00:00:00.000000  28435943.0   

       new_vaccinations  
0                   NaN  
1                   NaN  
2                   NaN  
3                   NaN  
4                   NaN  
...                 ...  
94000               NaN  
94001               NaN  
94002               NaN  
94003               NaN  
94004               NaN  

[94005 rows x 5 columns]

### Partitioning data with location and adding new vaccinations in a new coulmn by date

In [11]:
ps.sqldf("""
    SELECT cd.continent, cd.location, cd.date, cd.population, cv.new_vaccinations, SUM(cv.new_vaccinations) OVER (PARTITION BY cd.location ORDER BY cd.location, cd.date) AS  EveryDayCountAdded
    FROM cd
    JOIN cv
    ON cd.location = cv.location AND
    cd.date = cv.date
    WHERE cd.continent is not null
    AND cv.new_vaccinations is not null
    ORDER BY 2,3
""")


continent     location                        date  population  \
0          Asia  Afghanistan  2021-05-27 00:00:00.000000  38928341.0   
1          Asia  Afghanistan  2021-06-03 00:00:00.000000  38928341.0   
2        Europe      Albania  2021-01-13 00:00:00.000000   2877800.0   
3        Europe      Albania  2021-01-14 00:00:00.000000   2877800.0   
4        Europe      Albania  2021-01-15 00:00:00.000000   2877800.0   
...         ...          ...                         ...         ...   
11936    Africa     Zimbabwe  2021-06-20 00:00:00.000000  14862927.0   
11937    Africa     Zimbabwe  2021-06-21 00:00:00.000000  14862927.0   
11938    Africa     Zimbabwe  2021-06-22 00:00:00.000000  14862927.0   
11939    Africa     Zimbabwe  2021-06-23 00:00:00.000000  14862927.0   
11940    Africa     Zimbabwe  2021-06-24 00:00:00.000000  14862927.0   

       new_vaccinations  EveryDayCountAdded  
0                2859.0              2859.0  
1                4015.0              6874.0  
2                  60.0                60.0  
3                  78.0               138.0  
4                  42.0               180.0  
...                 ...                 ...  
11936            4813.0           1137419.0  
11937            2119.0           1139538.0  
11938            5526.0           1145064.0  
11939           20471.0           1165535.0  
11940           26466.0           1192001.0  

[11941 rows x 6 columns]

# Finding vaccinations for India with the above query
Vaccination in india starts on 16-Jan-2021

In [13]:
ps.sqldf("""
    SELECT cd.continent, cd.location, cd.date, cd.population, cv.new_vaccinations
    FROM cd
    JOIN cv
    ON cd.location = cv.location AND
    cd.date = cv.date
    WHERE cd.continent is not null 
    AND cd.location = 'India' 
    AND new_vaccinations is not null
    ORDER BY 1,2,3
""")


continent location                        date    population  \
0        Asia    India  2021-01-16 00:00:00.000000  1.380004e+09   
1        Asia    India  2021-01-17 00:00:00.000000  1.380004e+09   
2        Asia    India  2021-01-18 00:00:00.000000  1.380004e+09   
3        Asia    India  2021-01-19 00:00:00.000000  1.380004e+09   
4        Asia    India  2021-01-20 00:00:00.000000  1.380004e+09   
..        ...      ...                         ...           ...   
149      Asia    India  2021-06-23 00:00:00.000000  1.380004e+09   
150      Asia    India  2021-06-24 00:00:00.000000  1.380004e+09   
151      Asia    India  2021-06-25 00:00:00.000000  1.380004e+09   
152      Asia    India  2021-06-26 00:00:00.000000  1.380004e+09   
153      Asia    India  2021-06-27 00:00:00.000000  1.380004e+09   

     new_vaccinations  
0            191181.0  
1             33120.0  
2            229748.0  
3            220786.0  
4            131649.0  
..                ...  
149         3514882.0  
150         6683689.0  
151         6607501.0  
152         8179856.0  
153         3189028.0  

[154 rows x 5 columns]

### Finding total vaccinations, in addition and respect to each day

In [16]:
ps.sqldf("""
    SELECT cd.continent, cd.location, cd.date, cd.population, cv.new_vaccinations, SUM(cv.new_vaccinations) OVER (PARTITION BY cd.location ORDER BY cd.location, cd.date) AS Adding_Vaccinactions_on_days
    FROM cd
    JOIN cv
    ON cd.location = cv.location AND
    cd.date = cv.date
    WHERE cd.continent is not null
    AND cv.new_vaccinations is not null
    AND cd.location = 'India'
    ORDER BY 2,3
""")


continent location                        date    population  \
0        Asia    India  2021-01-16 00:00:00.000000  1.380004e+09   
1        Asia    India  2021-01-17 00:00:00.000000  1.380004e+09   
2        Asia    India  2021-01-18 00:00:00.000000  1.380004e+09   
3        Asia    India  2021-01-19 00:00:00.000000  1.380004e+09   
4        Asia    India  2021-01-20 00:00:00.000000  1.380004e+09   
..        ...      ...                         ...           ...   
149      Asia    India  2021-06-23 00:00:00.000000  1.380004e+09   
150      Asia    India  2021-06-24 00:00:00.000000  1.380004e+09   
151      Asia    India  2021-06-25 00:00:00.000000  1.380004e+09   
152      Asia    India  2021-06-26 00:00:00.000000  1.380004e+09   
153      Asia    India  2021-06-27 00:00:00.000000  1.380004e+09   

     new_vaccinations  Adding_Vaccinactions_on_days  
0            191181.0                      191181.0  
1             33120.0                      224301.0  
2            229748.0                      454049.0  
3            220786.0                      674835.0  
4            131649.0                      806484.0  
..                ...                           ...  
149         3514882.0                   281002852.0  
150         6683689.0                   287686541.0  
151         6607501.0                   294294042.0  
152         8179856.0                   302473898.0  
153         3189028.0                   305662926.0  

[154 rows x 6 columns]

In [ ]:
ps.sqldf("""
    SELECT cd.continent, cd.location, cd.date, cd.population, cv.new_vaccinations, (SUM(cv.new_vaccinations) OVER (PARTITION BY cd.location ORDER BY cd.location, cd.date) / cd.population) * 100 AS "Vaccinated %"
    FROM cd
    JOIN cv
    ON cd.location = cv.location AND
    cd.date = cv.date
    WHERE cd.continent is not null
    AND cv.new_vaccinations is not null
    AND cd.location = 'India'
    ORDER BY 2,3 DESC
    LIMIT 1
""")


### maximum vaccinations on a day

In [15]:
ps.sqldf("""
    SELECT cd.continent, cd.location, cd.date, cd.population, cv.new_vaccinations, MAX(cv.new_vaccinations) AS Max
    FROM cd
    JOIN cv
    ON cd.location = cv.location AND
    cd.date = cv.date
    WHERE cd.continent is not null
    AND cv.new_vaccinations is not null
    AND cd.location = 'India'
    ORDER BY 2,3
""")


continent location                        date    population  \
0      Asia    India  2021-06-21 00:00:00.000000  1.380004e+09   

   new_vaccinations        Max  
0         9027343.0  9027343.0

### Minimum vaccinations on a day

In [17]:
ps.sqldf("""
    SELECT cd.continent, cd.location, cd.date, cd.population, cv.new_vaccinations, MIN(cv.new_vaccinations) AS Min
    FROM cd
    JOIN cv
    ON cd.location = cv.location AND
    cd.date = cv.date
    WHERE cd.continent is not null
    AND cv.new_vaccinations is not null
    AND cd.location = 'India'
    ORDER BY 2,3
""")


continent location                        date    population  \
0      Asia    India  2021-01-26 00:00:00.000000  1.380004e+09   

   new_vaccinations     Min  
0            5671.0  5671.0

### From above queries on India
- Vaccination in india started on 16-Jan-2021
- Untill 27-June-2021 total vaccination is (305662926.0)
- On 21-June-21 Maximum number of vaccinations were done.(9027343)
- On 26-Jan-2021 Minimum number of vaccinations were done. (5671) (Republic Day)

## % of population vaccinated

In [12]:
ps.sqldf("""
With PopulationVsVaccination (continent, location, date, population, new_vaccinations, EveryDayCountAdded)
AS
(
    SELECT cd.continent, cd.location, cd.date, cd.population, cv.new_vaccinations, SUM(cv.new_vaccinations) OVER (PARTITION BY cd.location ORDER BY cd.location, cd.date) AS  EveryDayCountAdded
    FROM cd
    JOIN cv
    ON cd.location = cv.location AND
    cd.date = cv.date
    WHERE cd.continent is not null
    AND cv.new_vaccinations is not null
    ORDER BY 2,3
)
SELECT *, (EveryDayCountAdded / population ) * 100
FROM PopulationVsVaccination
""")


continent     location                        date  population  \
0          Asia  Afghanistan  2021-05-27 00:00:00.000000  38928341.0   
1          Asia  Afghanistan  2021-06-03 00:00:00.000000  38928341.0   
2        Europe      Albania  2021-01-13 00:00:00.000000   2877800.0   
3        Europe      Albania  2021-01-14 00:00:00.000000   2877800.0   
4        Europe      Albania  2021-01-15 00:00:00.000000   2877800.0   
...         ...          ...                         ...         ...   
11936    Africa     Zimbabwe  2021-06-20 00:00:00.000000  14862927.0   
11937    Africa     Zimbabwe  2021-06-21 00:00:00.000000  14862927.0   
11938    Africa     Zimbabwe  2021-06-22 00:00:00.000000  14862927.0   
11939    Africa     Zimbabwe  2021-06-23 00:00:00.000000  14862927.0   
11940    Africa     Zimbabwe  2021-06-24 00:00:00.000000  14862927.0   

       new_vaccinations  EveryDayCountAdded  \
0                2859.0              2859.0   
1                4015.0              6874.0   
2                  60.0                60.0   
3                  78.0               138.0   
4                  42.0               180.0   
...                 ...                 ...   
11936            4813.0           1137419.0   
11937            2119.0           1139538.0   
11938            5526.0           1145064.0   
11939           20471.0           1165535.0   
11940           26466.0           1192001.0   

       (EveryDayCountAdded / population ) * 100  
0                                      0.007344  
1                                      0.017658  
2                                      0.002085  
3                                      0.004795  
4                                      0.006255  
...                                         ...  
11936                                  7.652725  
11937                                  7.666982  
11938                                  7.704162  
11939                                  7.841894  
11940                                  8.019961  

[11941 rows x 7 columns]

In [14]:
ps.sqldf("""
With PopulationVsVaccination (continent, location, date, population, new_vaccinations, EveryDayCountAdded)
AS
(
    SELECT cd.continent, cd.location, cd.date, cd.population, cv.new_vaccinations, SUM(cv.new_vaccinations) OVER (PARTITION BY cd.location ORDER BY cd.location, cd.date) AS  EveryDayCountAdded
    FROM cd
    JOIN cv
    ON cd.location = cv.location AND
    cd.date = cv.date
    WHERE cd.continent is not null 
    AND cv.new_vaccinations is not null
    ORDER BY 2,3
)
SELECT *, (EveryDayCountAdded / population ) * 100
FROM PopulationVsVaccination
WHERE location = 'India'
""")


continent location                        date    population  \
0        Asia    India  2021-01-16 00:00:00.000000  1.380004e+09   
1        Asia    India  2021-01-17 00:00:00.000000  1.380004e+09   
2        Asia    India  2021-01-18 00:00:00.000000  1.380004e+09   
3        Asia    India  2021-01-19 00:00:00.000000  1.380004e+09   
4        Asia    India  2021-01-20 00:00:00.000000  1.380004e+09   
..        ...      ...                         ...           ...   
149      Asia    India  2021-06-23 00:00:00.000000  1.380004e+09   
150      Asia    India  2021-06-24 00:00:00.000000  1.380004e+09   
151      Asia    India  2021-06-25 00:00:00.000000  1.380004e+09   
152      Asia    India  2021-06-26 00:00:00.000000  1.380004e+09   
153      Asia    India  2021-06-27 00:00:00.000000  1.380004e+09   

     new_vaccinations  EveryDayCountAdded  \
0            191181.0            191181.0   
1             33120.0            224301.0   
2            229748.0            454049.0   
3            220786.0            674835.0   
4            131649.0            806484.0   
..                ...                 ...   
149         3514882.0         281002852.0   
150         6683689.0         287686541.0   
151         6607501.0         294294042.0   
152         8179856.0         302473898.0   
153         3189028.0         305662926.0   

     (EveryDayCountAdded / population ) * 100  
0                                    0.013854  
1                                    0.016254  
2                                    0.032902  
3                                    0.048901  
4                                    0.058441  
..                                        ...  
149                                 20.362461  
150                                 20.846785  
151                                 21.325587  
152                                 21.918329  
153                                 22.149417  

[154 rows x 7 columns]

### Total 22% of total population of India is now vaccinated (till 27th June) (first dose at least)

In [15]:
ps.sqldf("""
With PopulationVsVaccination (continent, location, date, population, new_vaccinations, EveryDayCountAdded)
AS
(
    SELECT cd.continent, cd.location, cd.date, cd.population, cv.new_vaccinations, SUM(cv.new_vaccinations) OVER (PARTITION BY cd.location ORDER BY cd.location, cd.date) AS  EveryDayCountAdded
    FROM cd
    JOIN cv
    ON cd.location = cv.location AND
    cd.date = cv.date
    WHERE cd.continent is not null 
    AND cv.new_vaccinations is not null
    ORDER BY 2,3
)
SELECT *, (EveryDayCountAdded / population ) * 100
FROM PopulationVsVaccination
WHERE location = 'United States'
""")


continent       location                        date   population  \
0    North America  United States  2020-12-21 00:00:00.000000  331002647.0   
1    North America  United States  2021-01-05 00:00:00.000000  331002647.0   
2    North America  United States  2021-01-06 00:00:00.000000  331002647.0   
3    North America  United States  2021-01-07 00:00:00.000000  331002647.0   
4    North America  United States  2021-01-08 00:00:00.000000  331002647.0   
..             ...            ...                         ...          ...   
159  North America  United States  2021-06-23 00:00:00.000000  331002647.0   
160  North America  United States  2021-06-24 00:00:00.000000  331002647.0   
161  North America  United States  2021-06-25 00:00:00.000000  331002647.0   
162  North America  United States  2021-06-26 00:00:00.000000  331002647.0   
163  North America  United States  2021-06-27 00:00:00.000000  331002647.0   

     new_vaccinations  EveryDayCountAdded  \
0             57909.0             57909.0   
1            273209.0            331118.0   
2            470328.0            801446.0   
3            612621.0           1414067.0   
4            768813.0           2182880.0   
..                ...                 ...   
159          648209.0         305827805.0   
160          815152.0         306642957.0   
161          512174.0         307155131.0   
162          923724.0         308078855.0   
163         1204225.0         309283080.0   

     (EveryDayCountAdded / population ) * 100  
0                                    0.017495  
1                                    0.100035  
2                                    0.242127  
3                                    0.427207  
4                                    0.659475  
..                                        ...  
159                                 92.394368  
160                                 92.640636  
161                                 92.795370  
162                                 93.074438  
163                                 93.438250  

[164 rows x 7 columns]

### Over 93% of USA population is now vaccinated (first Dose at least)